### PR TITLE
Vectorize the numeric part of Earliest Aggregator

### PR DESCRIPTION
**Pull request by somu-imply**
Vectorizing the earliest aggregator
```Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlExpressionBenchmark.querySql       42           5000000        false  avgt    5   86.661 ± 10.316  ms/op
SqlExpressionBenchmark.querySql       42           5000000        force  avgt    5   58.923 ±  9.156  ms/op
SqlExpressionBenchmark.querySql       43           5000000        false  avgt    5   76.705 ± 13.739  ms/op
SqlExpressionBenchmark.querySql       43           5000000        force  avgt    5   43.493 ±  7.034  ms/op
SqlExpressionBenchmark.querySql       44           5000000        false  avgt    5   75.060 ±  5.100  ms/op
SqlExpressionBenchmark.querySql       44           5000000        force  avgt    5   38.841 ±  1.883  ms/op
SqlExpressionBenchmark.querySql       45           5000000        false  avgt    5  189.588 ±  6.205  ms/op
SqlExpressionBenchmark.querySql       45           5000000        force  avgt    5   62.001 ±  9.902  ms/op
```

where
```
42 --> long
43 --> float
44 --> double
45 --> all three
```

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/query/aggregation/first/LongFirstVectorAggregator.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+
+/**
+ * Vectorized version of on heap 'last' aggregator for column selectors with type LONG..
+ */
+public class LongFirstVectorAggregator extends NumericFirstVectorAggregator
+{
+  long firstValue;
```
Comment (TSFenwick) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/LongFirstVectorAggregator.java:33]: private?

File: processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  @Override
+  public void aggregate(ByteBuffer buf, int position, int startRow, int endRow)
+  {
+    final long[] timeVector = timeSelector.getLongVector();
+    final boolean[] nullValueVector = valueSelector.getNullVector();
+    boolean nullAbsent = false;
+    firstTime = buf.getLong(position);
+    //check if nullVector is found or not
```
Comment (TSFenwick) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java:60]: nit: spacing on comments are inconsistent

File: processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+          updateTimeWithValue(buf, position, timeVector[row], row);
+        } else {
+          updateTimeWithNull(buf, position, timeVector[row]);
+        }
+      }
+    }
+  }
+
+  /**
+   *
```
Comment (TSFenwick) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java:131]: would be good to have an overall description of the method  be in the javadocs.

File: processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+   */
+  void updateTimeWithValue(ByteBuffer buf, int position, long time, int index)
+  {
+    buf.putLong(position, time);
+    buf.put(position + NULL_OFFSET, NullHandling.IS_NOT_NULL_BYTE);
+    putValue(buf, position + VALUE_OFFSET, index);
+  }
+
+  /**
+   *
```
Comment (TSFenwick) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java:145]: would be good to have an overall description of the method  be in the javadocs.

File: processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstVectorAggregator.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+import org.apache.druid.collections.SerializablePair;
+import org.apache.druid.segment.vector.VectorValueSelector;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+
+/**
+ * Vectorized version of on heap 'last' aggregator for column selectors with type LONG..
+ */
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstVectorAggregator.java:30]: stale javadocs. Similar comment for other `VectorAggregator` classes

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java [hunk 6]
```diff
@@ -995,6 +993,40 @@ public void testPrimitiveEarliestInSubquery() throws Exception
     );
   }
 
+  @Test
+  public void testPrimitiveEarliestInSubqueryGroupBy() throws Exception
```
Comment (suneet-s) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java:997]: This doesn't look like a subquery that is being tested

**Review by suneet-s — COMMENTED** [review]
```
|   0% (0/7)     |   0% (0/4)     |  73% (11/15)   | org/apache/druid/query/aggregation/first/FloatFirstAggregatorFactory.java
```

Overall LGTM. This is the only thing that slightly concerns me. Can you look into why there's no coverage for changes to this class, while there is coverage for the other aggregator factories.

**Comment by somu-imply** [comment]
@suneet-s I have added additional tests that would cover the factory classes for all the three new additions. The coverage would improve now.
@TSFenwick I have updated as per your review comments.

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java (hunk 6 above)
Comment (somu-imply) [reply sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java]: updated

File: processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstVectorAggregator.java (hunk 5 above)
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstVectorAggregator.java:30]: updated

File: processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java (hunk 4 above)
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java:150]: done

File: processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java (hunk 3 above)
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java:135]: done

File: processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java (hunk 2 above)
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java]: adjusted

**Comment by suneet-s** [comment]
@somu-imply It looks like there are still some failing tests, can you fix them up please.

`TimeseriesQueryRunnerTest.testTimeseriesWithFirstLastAggregator` is one example

**Comment by somu-imply** [comment]
Instead of conflict resolution that's gonna make things difficult, I'll make a fresh PR will close this once the new one's up and linked to this

**Pull request closed without merging**